source/ui/DisplayWindow.ooc

```diff
@@ -15,15 +15,60 @@
  * along with this software. If not, see <http://www.gnu.org/licenses/>.
  */
 
+use ooc-base
 use ooc-draw
 use ooc-geometry
 import x11/UnixWindow
 import win32/Win32DisplayWindow
+import GuiEvent
 
 DisplayWindow: abstract class {
+	_mousePressHandler: Event1<MouseEvent>
+	_mouseReleaseHandler: Event1<MouseEvent>
+	_keyPressHandler: Event1<KeyboardEvent>
+	_keyReleaseHandler: Event1<KeyboardEvent>
 	init: func (size: IntVector2D, title: String)
+	free: override func {
+		if (this _mousePressHandler)
+			this _mousePressHandler free()
+		if (this _mouseReleaseHandler)
+			this _mouseReleaseHandler free()
+		if (this _keyPressHandler)
+			this _keyPressHandler free()
+		if (this _keyReleaseHandler)
+			this _keyReleaseHandler free()
+		super()
+	}
 	draw: abstract func (image: Image)
 	refresh: virtual func
+	addMousePressHandler: func (handler: Event1<MouseEvent>) {
+		this _mousePressHandler = this _mousePressHandler ? this _mousePressHandler add(handler) : handler
```

Comment:
We removed it some time ago as it was not working in some cases, and we wanted to have single interface in all cases.